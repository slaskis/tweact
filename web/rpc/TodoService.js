import { withTwirp, TwirpService } from "./twirp";

export const ListTodos = withTwirp(
  class ListTodos extends TwirpService {
    constructor(props) {
      super("todos.v1.TodoService/ListTodos", props);
    }
  }
);

export const CreateTodo = withTwirp(
  class CreateTodo extends TwirpService {
    constructor(props) {
      super("todos.v1.TodoService/CreateTodo", props);
    }
  }
);
