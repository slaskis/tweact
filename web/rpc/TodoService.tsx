import { withTwirp, TwirpService } from "./twirp";

// TODO don't export these, instead use a separate import (twirp/json, twirp/protobuf)
export {
  TwirpJSONClient as TwirpClient,
  TwirpProvider,
  InMemoryCache,
  renderState
} from "./twirp";

export interface Todo {
  id: string;
  title: string;
}

export interface TodoResponse {
  todo?: Todo;
}

export interface CreateTodoRequest {
  title: string;
}
export interface ListTodoRequest {}

export interface ListTodoResponse {
  todos?: Todo[];
}

export const ListTodos = withTwirp<ListTodoRequest, ListTodoResponse>(
  class ListTodos extends TwirpService<ListTodoRequest, ListTodoResponse> {
    constructor(props: any) {
      super("todos.v1.TodoService/ListTodos", props);
    }
  }
);

export const CreateTodo = withTwirp<CreateTodoRequest, TodoResponse>(
  class CreateTodo extends TwirpService<CreateTodoRequest, TodoResponse> {
    constructor(props: any) {
      super("todos.v1.TodoService/CreateTodo", props);
    }
  }
);
