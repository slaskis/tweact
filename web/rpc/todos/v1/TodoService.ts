// generated by protoc-gen-tweact. DO NOT EDIT.
// source: todos/v1/service.proto

import { TwirpClient } from "@department/twirp-component";

export interface Todo {
  id: string;
  title: string;
  state: TodoState;
}
export interface TodoResponse {
  todo: Todo;
}
export interface ListTodoResponse {
  todos: Todo[];
}
export interface CreateTodoRequest {
  title: string;
}
export interface RemoveTodoRequest {
  id: string;
}
export interface GetTodoRequest {
  id: string;
}
export interface ListTodosRequest {}
export enum TodoState {
  UNKNOWN = 0,
  ARCHIVED = 1,
  ACTIVE = 2
}

export const CreateTodo = (
  r: CreateTodoRequest,
  t: TwirpClient<CreateTodoRequest, TodoResponse>
) => t.request("todos.v1.TodoService/CreateTodo", r, {});
export const RemoveTodo = (
  r: RemoveTodoRequest,
  t: TwirpClient<RemoveTodoRequest, TodoResponse>
) => t.request("todos.v1.TodoService/RemoveTodo", r, {});
export const GetTodo = (
  r: GetTodoRequest,
  t: TwirpClient<GetTodoRequest, TodoResponse>
) => t.request("todos.v1.TodoService/GetTodo", r, {});
export const ListTodos = (
  r: ListTodosRequest,
  t: TwirpClient<ListTodosRequest, ListTodoResponse>
) => t.request("todos.v1.TodoService/ListTodos", r, {});
