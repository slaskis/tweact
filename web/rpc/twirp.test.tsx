import React from "react";
import renderer from "react-test-renderer";
import {
  TwirpJSONClient,
  renderState,
  TwirpProvider,
  InMemoryCache
} from "./twirp";
import { ListTodos } from "./TodoService";

test("none", async () => {
  const cache = new InMemoryCache();
  const client = new TwirpJSONClient("http://localhost:4000/twirp/", cache);
  const App = () => <div />;
  await renderState(client, <App />);
  expect(cache.store.size).toBe(0);
  expect(renderer.create(<App />).toJSON()).toMatchSnapshot();
});

test("client in context", async () => {
  const cache = new InMemoryCache();
  const client = new TwirpJSONClient("http://localhost:4000/twirp/", cache);

  const spy = jest.spyOn(client, "request");
  const App = () => (
    <ListTodos>{({ loading }) => (loading ? "LOADING" : "OK")}</ListTodos>
  );
  await renderState(client, <App />);
  expect(cache.store.size).toBe(1);
  expect(spy).toHaveBeenCalledTimes(2); // once with pending, once without
  spy.mockReset();
  expect(
    renderer
      .create(
        <TwirpProvider value={{ client }}>
          <App />
        </TwirpProvider>
      )
      .toJSON()
  ).toMatchSnapshot();
  expect(cache.store.size).toBe(1);
  expect(spy).toHaveBeenCalledTimes(1); // should be cached now
});
