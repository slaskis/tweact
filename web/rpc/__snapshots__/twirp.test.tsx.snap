// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client in context 1`] = `"OK"`;

exports[`client in context 2`] = `"LOADING"`;

exports[`none 1`] = `<div />`;
